src/H5Dmpio.c

```diff
@@ -2927,187 +3134,352 @@ H5D__obtain_mpio_mode(H5D_io_info_t *io_info, H5D_dset_io_info_t *di, uint8_t as
+                 * extensible array code calculated instead of what was calculated
+                 * in the chunk file mapping.
+                 */
+                if (di[dset_idx].dset->shared->layout.u.chunk.idx_type == H5D_CHUNK_IDX_EARRAY)
+                    local_info_array[buf_idx].index_info.chunk_idx = udata.chunk_idx;
+                else
+                    local_info_array[buf_idx].index_info.chunk_idx = chunk_info->index;
+
+                assert(H5_addr_defined(di[dset_idx].dset->oloc.addr));
+                local_info_array[buf_idx].index_info.dset_oloc_addr = di[dset_idx].dset->oloc.addr;
+
+                local_info_array[buf_idx].index_info.filter_mask = udata.filter_mask;
+                local_info_array[buf_idx].index_info.need_insert = false;
+
+                buf_idx++;
+
+                chunk_node = H5SL_next(chunk_node);
             }
+        }
+        else if (H5F_get_coll_metadata_reads(di[dset_idx].dset->oloc.file)) {
+            hsize_t scaled[H5O_LAYOUT_NDIMS] = {0};
 
             /*
-             * Extensible arrays may calculate a chunk's index a little differently
-             * than normal when the dataset's unlimited dimension is not the
-             * slowest-changing dimension, so set the index here based on what the
-             * extensible array code calculated instead of what was calculated
-             * in the chunk file mapping.
+             * If this rank has no selection in the dataset and collective
+             * metadata reads are enabled, do a fake lookup of a chunk to
+             * ensure that this rank has the chunk index opened. Otherwise,
+             * only the ranks that had a selection will have opened the
+             * chunk index and they will have done so independently. Therefore,
+             * when ranks with no selection participate in later collective
+             * metadata reads, they will try to open the chunk index collectively
+             * and issues will occur since other ranks won't participate.
+             *
+             * In the future, we should consider having a chunk index "open"
+             * callback that can be used to ensure collectivity between ranks
+             * in a more natural way, but this hack should suffice for now.
              */
-            if (di->dset->shared->layout.u.chunk.idx_type == H5D_CHUNK_IDX_EARRAY)
-                local_info_array[i].index_info.chunk_idx = udata.chunk_idx;
-            else
-                local_info_array[i].index_info.chunk_idx = chunk_info->index;
-
-            local_info_array[i].index_info.filter_mask = udata.filter_mask;
-            local_info_array[i].index_info.need_insert = false;
-
-            chunk_node = H5SL_next(chunk_node);
+            if (H5D__chunk_lookup(di[dset_idx].dset, scaled, &udata) < 0)
+                HGOTO_ERROR(H5E_DATASET, H5E_CANTGET, FAIL, "error looking up chunk address");
```

Comment:
This fake lookup should no longer be necessary with the changes to the chunk `io_init` callback, but is probably better addressed with the solution to #2658